main/lsp/request_dispatch.cc

```diff
@@ -21,8 +25,262 @@ class SorbetFenceTask final : public LSPTask {
+            initialGS->errorQueue->drainWithQueryResponses();
+        }
+
+        // When inputFileNames is 0 (as in tests), indexed ends up being size 0 because we don't index payload files.
+        // At the same time, we expect indexed to be the same size as GlobalStateHash, which _does_ have payload files.
+        // Resize the indexed array accordingly.
+        if (indexed.size() < initialGS->getFiles().size()) {
+            indexed.resize(initialGS->getFiles().size());
+        }
+
+        return indexed;
+    }
+
+public:
+    // Outputs:
+    // - initialGS that has indexed the workspace.
+    // - globalStateHashes for every file in workspace.
+    // - updates containing the state needed to pass to the typechecker for initialization.
+    IndexAndHashWorkspaceTask(const LSPConfiguration &config, unique_ptr<core::GlobalState> &initialGS,
+                              vector<core::FileHash> &globalStateHashes, LSPFileUpdates &updates)
+        : LSPTask(config, true), initialGS(initialGS), globalStateHashes(globalStateHashes), updates(updates) {}
+
+    void run(LSPTypecheckerDelegate &tc) override {
+        auto &workers = tc.workers;
+        // Temporarily replace error queue, as it asserts that the same thread that created it uses it and we're
+        // going to use it on typechecker thread for this one operation.
+        auto savedErrorQueue = initialGS->errorQueue;
+        initialGS->errorQueue = make_shared<core::ErrorQueue>(savedErrorQueue->logger, savedErrorQueue->tracer);
+        // Enforce that this is only run once.
+        ENFORCE(globalStateHashes.empty());
+
+        vector<ast::ParsedFile> indexed;
+        {
+            Timer timeit(config.logger, "initial_index");
+            ShowOperation op(config, "Indexing", "Indexing files...");
+            indexed = indexFromFileSystem(workers, nullptr /* TODO(jvilk): Thread through kvstore */);
+            globalStateHashes = LSPTypechecker::computeFileHashes(config, initialGS->getFiles(), workers);
+        }
+        updates.epoch = 0;
+        updates.canTakeFastPath = false;
+        // *Copy* global state hashes; both LSPLoop and LSPTypechecker need a copy (LSPLoop to figure out
+        // cancelation, LSPTypechecker to run queries)
+        updates.updatedFileHashes = globalStateHashes;
+        updates.updatedFileIndexes = move(indexed);
+        updates.updatedGS = initialGS->deepCopy();
+
+        // Restore error queue, as initialGS will be used on the LSPLoop thread from now on.
+        initialGS->errorQueue = move(savedErrorQueue);
+    }
+};
+
+const core::FileHash &findHash(int id, const vector<core::FileHash> &globalStateHashes,
+                               const UnorderedMap<int, core::FileHash> &overriddingStateHashes) {
+    const auto it = overriddingStateHashes.find(id);
+    if (it == overriddingStateHashes.end()) {
+        return globalStateHashes[id];
+    }
+    return it->second;
+}
+
+bool canTakeFastPath(
```

Comment:
Mostly unchanged from `TimeTravelingGlobalState`, but introduces the notion of `overriddingStateHashes`.

Three important things to know:
1. These file hashes are what we use to determine if edits can take the fast or slow path.
2. `globalStateHashes` contains the latest file hashes as of this update.
3. If a slow path is running, we need to be able to tell if an edit would take the fast path with the edits typechecking on the slow path.

In order to accomplish 3 given 1 and 2, `overridingStateHashes` contains the file hashes evicted from `globalStateHashes` since the slow path began. It lets us make a fast path decision in the immediate past.